### ISSUE TITLE
Connection or File Descriptor leak when HTTP Exception occurs in async batch

### ISSUE THREAD

**Issue by michaelbutler** [guzzle/guzzle#2289]
**Guzzle version(s) affected**: 6.3.0 and possibly older versions

**Description**  
When using the async feature to make a batch of requests in parallel, if one of those requests responds back with an exception type response (example, one of them is a 404 and the rest are normal 200) there will be an open file descriptor somewhere in the stack -- e.g. fclose is not being closed properly on something -- with no way of cleaning up this open handle except by quitting the process (not valid for long running PHP scripts).

**How to reproduce**  
Reproduce is simple using a fairly simple script, occurs on Linux based system but also macOS with PHP 7.1.

0. Have PHP 7.1 installed with cURL available, on a linux or macOS system
1. Download [gist](https://gist.github.com/michaelbutler/4800e7f5135acfc2f2c4166facd3272f) to an empty dir and name it `gist.php`
2. Run `composer require guzzlehttp/guzzle` to get latest guzzle installed into `vendor/`
3. Run `php gist.php`

**Expected:**
Because the inner code operates in a function with only local variables, we expect that as execution exits the function, all objects that were created would have their `__destruct()` called, freeing up any open handles that were generated.

**Actual:**
As the script runs and the URLs are hit in parallel, the number of open file descriptors (labeled as OPEN FDs in the debug output) increases for every iteration (the main loop runs 4 times total). For bigger running batch scripts you could imagine that eventually the "open file limit" will be hit due to HTTP exceptions and bad URLs occurring.

**Inverse case**
To observe what would happen in a perfect world where file descriptors are NOT kept open, we can make sure all URLs respond with a 200:
1. Comment out using `//` on **line 51**, save and re-run the test script
2. Notice that since all URLs are 200 and no exceptions occur, the file descriptors (OPEN FDs) never increase indefinitely, they'll stay as some static number.

**Possible Solution**  
Likely something to do with the way CurlHandler, CurlMultiHandler, or CurlFactory works. However it could be an internal PHP or CURL bug. Could also be the Guzzle Promises implementation.

**Additional context**

**Comment by michaelbutler** [comment]
a workaround was found:

I have finally gotten to the bottom of this and have a workaround. It involves creating a CurlFactory class override in your own repository that extends Guzzle's CurlFactory. A custom method `cleanup` is added that you have to call periodically -- ideally when your inner code is completely finished making requests and using the response(s). In my testing, this never increases the open file descriptor count and also seems to be memory efficient and also fast (does requests in parallel asynchronously)

See this [updated gist](https://github.tumblr.net/gist/mbutler/06d94089e391d3343161cc2b2dd4fe40) file which includes the example class override.

I believe the core problem is that `curl_close` is never called on the handles in CurlFactory even if your own functions return and their local variables not needed anymore. By adding the `cleanup` method we loop through and manually close these open connections.

I'm not sure yet how to cleanly work this into the internal logic of Guzzle so that everyone can take advantage of it (patching the leak) but maybe someone can.

**Comment by arderyp** [comment]
thanks @michaelbutler.  I want to test this, but your gist link is broken

**Comment by arderyp** [comment]
I'm guessing this is what you meant to post: https://gist.github.com/michaelbutler/14195fbf91a525c5af0cdcbf819095eb

**Comment by arderyp** [comment]
@michaelbutler , sadly, I don't think your proposed solution works.  It creates the same problem I described in [this comment](https://github.com/guzzle/guzzle/issues/1927#issuecomment-431467861).  That is, it breaks concurrency.

When I use:

```
$maxHandles = 50;
$handlerFactory = new  MyCurlFactory($maxHandles);
$handlerOptions = ['handle_factory' => $handlerFactory];
$handler = new CurlHandler($handlerOptions);

$clientOptions = [
    'handler' => HandlerStack::create($handler),
    'proxy' => null,
];
$client = new Client($clientOptions);
```

it simply executes all requests linearly and takes forever.  And, as I mentioned in the aforementioned comment, the `$maxHandles` value is completely worthless (I think that's a separate bug).  No matter what number I give it, it always runs the same speed in the same fashion.  Your approach does avoid the maximum open handlers error, but so did the problematic approach I described in the aforementioned comment, and that approach didn't require creating a custom class to override CurlFactory.

Am I missing something?

**Comment by arderyp** [comment]
The speed issue doesn't seem to be a problem with your demo request script, so I am not sure if maybe there is some other aspect of my implementation that is causing your approach and the other to break my concurrency.

But, while your new demo does cause the number of open FD handlers to be much reduced from your previous demo, the FD count does still appear to be going up endlessly (from 5 to 6 to 7 to 8 to 9):

```
*** IN BEGIN: OPEN FDs 5
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 6
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 7
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 8
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 9
*** IN FINAL: OPEN FDs 9
```

**Comment by michaelbutler** [comment]
That's quite odd you're seeing that. It almost looks exactly like the previous version of the repro script without the fix. I just redownloaded the two gist files and got this (macOS with PHP 7.1 installed via brew)

```
*** IN BEGIN: OPEN FDs 42
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 42
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 42
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 42
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 42
*** IN FINAL: OPEN FDs 42
```

You can see the FDs don't increase even though the 404 is happening.

Also just re-ran it on a Linux based system and got:

```
*** IN BEGIN: OPEN FDs 6
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 8
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 8
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 8
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got exception: Client error: `GET https://66.media.tumblr.com/c7e0b3c2266731194761a2834c/tumblr_pjyk879oq91qdj91io2_540.png` resulted in a `404 Not Found` response:
404 Not Found


Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
Got response! Code 200
*** IN LOOP: OPEN FDs 8
*** IN FINAL: OPEN FDs 8
```

can you confirm you're in a new project with Guzzle 6.3.3 installed from scratch? (composer require guzzlehttp/guzzle)

**Comment by michaelbutler** [comment]
Regarding your own implementation I'd need to look at how you're doing the requests themselves (the part where async requests are created and then the promises are settled/waited on)

**Comment by arderyp** [comment]
Thanks @michaelbutler , I will comment again when I have a change to reinstall your gist+guzzle from scratch and provide details on my implementation

**Comment by arderyp** [comment]
@michaelbutler, on re-installing your gist, the output was as expected, with no incrementing FDs.

So, on to my implementation.

Here's a dump of the config for the client object in your gist:

```
array(9) {
  'base_uri' =>
  class GuzzleHttp\Psr7\Uri#10 (7) {
    private $scheme =>
    string(5) "https"
    private $userInfo =>
    string(0) ""
    private $host =>
    string(19) "66.media.tumblr.com"
    private $port =>
    NULL
    private $path =>
    string(1) "/"
    private $query =>
    string(0) ""
    private $fragment =>
    string(0) ""
  }
  'proxy' =>
  NULL
  'handler' =>
  class GuzzleHttp\HandlerStack#5 (3) {
    private $handler =>
    class GuzzleHttp\Handler\CurlHandler#2 (1) {
      private $factory =>
      class MyCurlFactory#3 (3) {
        ...
      }
    }
    private $stack =>
    array(4) {
      [0] =>
      array(2) {
        ...
      }
      [1] =>
      array(2) {
        ...
      }
      [2] =>
      array(2) {
        ...
      }
      [3] =>
      array(2) {
        ...
      }
    }
    private $cached =>
    NULL
  }
  'allow_redirects' =>
  array(5) {
    'max' =>
    int(5)
    'protocols' =>
    array(2) {
      [0] =>
      string(4) "http"
      [1] =>
      string(5) "https"
    }
    'strict' =>
    bool(false)
    'referer' =>
    bool(false)
    'track_redirects' =>
    bool(false)
  }
  'http_errors' =>
  bool(true)
  'decode_content' =>
  bool(true)
  'verify' =>
  bool(true)
  'cookies' =>
  bool(false)
  'headers' =>
  array(1) {
    'User-Agent' =>
    string(70) "GuzzleHttp/6.3.3 curl/7.47.0 PHP/7.2.17-1+ubuntu16.04.1+deb.sury.org+3"
  }
}
```

Here's a dump of the client config used in my implementation:

```
array(12) {
  'base_uri' =>
  class GuzzleHttp\Psr7\Uri#61 (7) {
    private $scheme =>
    string(5) "https"
    private $userInfo =>
    string(0) ""
    private $host =>
    string(17) "apis.berkeley.edu"
    private $port =>
    NULL
    private $path =>
    string(0) ""
    private $query =>
    string(0) ""
    private $fragment =>
    string(0) ""
  }
  'timeout' =>
  int(100)
  'headers' =>
  array(5) {
    'Accept' =>
    string(16) "application/json"
    'Connection' =>
    string(5) "close"
    'app_id' =>
    string(8) "c0b480d1"
    'app_key' =>
    string(32) "be6bb2952fcc6e5e1c4692e676c71bba"
    'User-Agent' =>
    string(70) "GuzzleHttp/6.3.3 curl/7.47.0 PHP/7.2.17-1+ubuntu16.04.1+deb.sury.org+3"
  }
  [75] =>
  bool(true)
  [74] =>
  bool(true)
  'handler' =>
  class GuzzleHttp\HandlerStack#55 (3) {
    private $handler =>
    class GuzzleHttp\Handler\CurlHandler#54 (1) {
      private $factory =>
      class My\Namespace\Root\MyCurlFactory#53 (3) {
        ...
      }
    }
    private $stack =>
    array(4) {
      [0] =>
      array(2) {
        ...
      }
      [1] =>
      array(2) {
        ...
      }
      [2] =>
      array(2) {
        ...
      }
      [3] =>
      array(2) {
        ...
      }
    }
    private $cached =>
    NULL
  }
  'proxy' =>
  NULL
  'allow_redirects' =>
  array(5) {
    'max' =>
    int(5)
    'protocols' =>
    array(2) {
      [0] =>
      string(4) "http"
      [1] =>
      string(5) "https"
    }
    'strict' =>
    bool(false)
    'referer' =>
    bool(false)
    'track_redirects' =>
    bool(false)
  }
  'http_errors' =>
  bool(true)
  'decode_content' =>
  bool(true)
  'verify' =>
  bool(true)
  'cookies' =>
  bool(false)
}
```

As far as I can tell, they aren't significantly different.  I will post more details about my specific code implementation in the next comment.

**Comment by arderyp** [comment]
@michaelbutler, I am still working on replicating my failing use case in some sort of generic gist.  I predict it will be quite a complicated task.  I will continue to keep you posted.  For what it is worth, I am not re-instantiating the client and factory for each processing chunk/loop, like you are.  I have one primary object that instantiates a local client object, then generates thousands of promises based on calls I make on the object, then a separate object handles the unwrapping and mapping of those responses.

On a side note, I noticed that your new gist clearly demonstrates how the `max_handles` value passed into the `CurlHandler` doesn't do anything at all.  I've copy/pasted the list of paths in your script, essentially making the list twice as long, hitting each path you originally provided twice.  I did this just to have a larger set of requests to tests against.  And, no matter what `int` I pass into `MyCurlFactory`, the output always consistently reads `OPEN FDs 5`.  I also increase the high range number in the `iter` loop to execute more loops, and still the ouput always ready `OPEN FDs 5`.  So, I have no idea where that number is coming from or how to throttle it--it seems to me the intended `max_handles` argument to the handler is the intended method, but doesn't do anything.  Do you agree with this analysis?  If so, should we open another separate issue for that?

**Comment by michaelbutler** [comment]
@arderyp can you contact me via some other means instead of this Pull Request?

**Comment by michaelbutler** [comment]
@arderyp let's move this to https://gitter.im/guzzle/guzzle

**Comment by arderyp** [comment]
@michaelbutler I just emailed the address published on your website.

**Comment by gmponos** [comment]
Is this related with this? https://github.com/guzzle/promises/pull/109

**Comment by michaelbutler** [comment]
@gmponos Interesting! I'm nearly certain it is related! We can see if my leak repro script in this issue's description goes away when running code from that PR's branch.

**Comment by michaelbutler** [comment]
@gmponos Although it seems very related, I have re-tested this reproduction script on various guzzle changes since April 2019. I reproduced the issue again on the Guzzle version at the time (tested 6.3.2). Then I upgraded to GuzzleHttp 6.4.1 and the issue went away... so apparently my issue was fixed at some point since this issue was created (or in the dependent composer packages, e.g. guzzle/promises, during that time). 

Closing this issue now, Thanks.
